testopt/testopt.go

```diff
@@ -0,0 +1,132 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package testopt provides objective functions for testing optimization algorithms.
+package testopt
+
+import (
+	"math"
+	"reflect"
+	"testing"
+
+	"github.com/gonum/floats"
+)
+
+// Minimum represents information about an optimal location of a function.
+type Minimum struct {
+	// X is the location of the minimum. If it is unknown or if the location is
+	// not a single point, X is nil.
+	X []float64
+	// F is the value of the objective function at X.
+	F float64
+	// Global indicates whether the minimum is local or global.
+	Global bool
+}
+
+// Function represents a test objective function.
+type Function interface {
+	F(x []float64) float64
+
+	// Minima returns known minima of the function. If no minima exist or if
+	// their location for the given dimension is unknown, nil is returned.
+	Minima() []Minimum
+}
+
+type Gradient interface {
+	Df(x, grad []float64)
+}
+
+type FunctionGradient interface {
+	FDf(x, grad []float64) float64
+}
+
+const (
+	defaultTol     = 1e-8
+	defaultGradTol = 1e-4
+)
+
+func testFunction(f Function, tests []funcTest, t *testing.T) {
```

Comment:
In addition to (instead of?) wantGrad, maybe we should also test a couple of random points and compare with gonum/diff/fd.Gradient. For wantGrad, it's tempting to choose "simple" x locations like 0 and 1, which are the most likely to cause false positives. It's also a way to test the higher-dimensional functions where we have no interest in explicitly writing the gradient. fd.Gradient can be executed concurrently if we're worried about testing time.